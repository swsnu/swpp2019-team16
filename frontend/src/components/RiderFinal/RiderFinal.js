import React from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/styles';
import theme from '../../lib/styles/theme';

const RiderFinalStyles = makeStyles({
  root: {
    position: 'absolute',
    left: 0,
    top: 0,
    bottom: 0,
    right: 0,
    backgroundColor: theme.palette.primary.main,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
  },
  container: {
    marginTop: '20px',
    marginBottom: '20px',
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
  },
  whiteBox: {
    boxShadow: '0 0 8px rgba(0, 0, 0, 0.25)',
    padding: '2rem',
    width: '400px',
    background: 'white',
    borderRadius: '2px',
  },
  button: {
    background: theme.palette.primary.dark,
    borderRadius: 5,
    color: 'white',
    cursor: 'pointer',
    paddingLeft: '1rem',
    paddingRight: '1rem',
    paddingBottom: '0.3rem',
    paddingTop: '0.3rem',
    marginBottom: '10px',
    marginTop: '10px',
    marginLeft: '10px',
    marginRight: '10px',
  },
});

RiderFinal.propTypes = {
  group: PropTypes.object.isRequired,
};

function RiderFinal({ group, onClickGoToMain }) {
  const styles = RiderFinalStyles();
  const premiumTotalCost = group
    ? Math.floor((group.cost * 1.2) / 100) * 100
    : 0;
  const riderFee = group ? group.riderCost : 0;
  const saved = premiumTotalCost - riderFee;
  console.log('rider final component');
  return (
    <div className={styles.root}>
      <div className={styles.whiteBox}>
        <div className={styles.container}>
          <Typography variant={'body1'}>
            TotalCost is &#8361;{premiumTotalCost}
          </Typography>
          <Typography variant={'body1'}>Your fee &#8361;{riderFee}</Typography>
          <Typography variant={'body1'}>You saved &#8361;{saved}</Typography>
          <div
            className={styles.button}
            children="Go To Main"
            variant="contained"
            fullwidth="false"
            onClick={() => onClickGoToMain()}
          />
        </div>
      </div>
    </div>
  );
}
export default RiderFinal;
