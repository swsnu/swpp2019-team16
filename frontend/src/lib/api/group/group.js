import client from '../client';

export function createGroup({ groupId, riders, driver, from, to }) {
  return client.post(`/api/v1/group`, { groupId, riders, driver, from, to });
}

export function acceptGroup({ groupId, driverId }) {
  return client.put(`/api/v1/group/${groupId}`, driverId);
}

export function onTaxi(riderId) {
  return client.put(`/api/v1/group/taxi/riders`, riderId);
}

export function departure({ groupId, departureTime }) {
  return client.put(`/api/v1/group/taxi/departure`, { groupId, departureTime });
}

export function fetchMyGroup(groupId) {
  return client.get(`/api/v1/group/${groupId}`);
}

export function notifyDriverLocation({ driverId, groupId, locationInfo }) {
  return client.put(`/api/v1/group/taxi/location`, {
    driverId,
    groupId,
    locationInfo,
  });
}

export function confirmCost({ groupId, totalCost }) {
  return client.put(`/api/v1/group/cost`, { groupId, totalCost });
}
