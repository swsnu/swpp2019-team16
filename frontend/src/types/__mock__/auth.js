export const mockRiderAuth = {
  id: 1,
  email: 'rider1@gmail.com',
  user_type: 'rider',
  vehicle: {
    car_type: '',
    plate_no: '',
  },
};

export const mockDriverAuth = {
  id: 1,
  email: 'driver1@gmail.com',
  user_type: 'driver',
  vehicle: {
    car_type: '',
    plate_no: '',
  },
};
