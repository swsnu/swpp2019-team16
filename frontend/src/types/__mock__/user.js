export const mockRider = {
  id: 1, // rider/driver id
  status: 'IDLE',
  group: 1, // group id
  user: {
    id: 2,
    email: 'rider1@gmail.com',
    user_type: 'rider',
    vehicle: {
      car_type: '',
      plate_no: '',
    },
  },
};

export const mockDriver = {
  id: 1, // rider/driver id
  status: 'IDLE',
  group: 1, // group id
  user: {
    id: 2,
    email: 'driver1@gmail.com',
    user_type: 'driver',
    vehicle: {
      car_type: 'benz',
      plate_no: '1234',
    },
  },
};

export const createMockRider = (id, groupId, email) => {
  return {
    id,
    status: 'IDLE',
    group: groupId,
    user: {
      id: 2,
      email,
      user_type: 'rider',
      vehicle: {
        car_type: '',
        plate_no: '',
      },
    },
  };
};

export const createMockDriver = (id, groupId, email) => {
  return {
    id,
    status: 'IDLE',
    group: groupId,
    user: {
      id: 2,
      email,
      user_type: 'driver',
      vehicle: {
        car_type: 'benz',
        plate_no: '1234',
      },
    },
  };
};
