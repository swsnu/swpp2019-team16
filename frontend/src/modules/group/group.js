import { createAction, handleActions } from 'redux-actions';
import { takeLatest } from 'redux-saga/effects';
import * as groupAPI from '../../lib/api/group/group';
import createRequestSaga, {
  createRequestActionTypes,
} from '../../lib/createRequestSaga';

const [
  ACCEPT_GROUP,
  ACCEPT_GROUP_SUCCESS,
  ACCEPT_GROUP_FAILURE,
] = createRequestActionTypes('group/ACCEPT_GROUP');

const [ON_TAXI, ON_TAXI_SUCCESS, ON_TAXI_FAILURE] = createRequestActionTypes(
  'group/ON_TAXI',
);

const [
  DEPARTURE,
  DEPARTURE_SUCCESS,
  DEPARTURE_FAILURE,
] = createRequestActionTypes('group/DEPARTURE');

const [
  FETCH_MY_GROUP,
  FETCH_MY_GROUP_SUCCESS,
  FETCH_MY_GROUP_FAILURE,
] = createRequestActionTypes('group/FETCH_MY_GROUP');

const [
  NOTIFY_DRIVER_LOCATION,
  NOTIFY_DRIVER_LOCATION_SUCCESS,
  NOTIFY_DRIVER_LOCATION_FAILURE,
] = createRequestActionTypes('group/NOTIFY_DRIVER_LOCATION');

const [
  CONFIRM_COST,
  CONFIRM_COST_SUCCESS,
  CONFIRM_COST_FAILURE,
] = createRequestActionTypes('group/CONFIRM_COST');

export const GROUP_CREATED = 'group/GROUP_CREATED';
export const GROUP_UPDATED = 'group/GROUP_UPDATED';
export const UNLOAD_GROUP = 'group/UNLOAD_GROUP';
export const GROUP_COST_UPDATED = 'group/GROUP_COST_UPDATED';
export const INIT_COST_CONFIRMED = 'group/INIT_COST_CONFIRMED';

export const groupCreated = createAction(
  GROUP_CREATED,
  ({ groupId, from, to }) => ({
    groupId,
    from,
    to,
  }),
);

export const groupCostUpdated = createAction(
  GROUP_COST_UPDATED,
  ({ groupId, riderCost, cost }) => ({
    groupId,
    riderCost,
    cost,
  }),
);

export const groupUpdated = createAction(
  GROUP_UPDATED,
  ({ groupId, driver, riders, fromLocation, toLocation }) => ({
    groupId,
    driver,
    riders,
    from: fromLocation,
    to: toLocation,
  }),
);

export const unloadGroup = createAction(UNLOAD_GROUP);

export const acceptGroup = createAction(
  ACCEPT_GROUP,
  ({ groupId, driverId }) => ({
    groupId,
    driverId,
  }),
);
export const onTaxi = createAction(ON_TAXI, riderId => riderId);
export const departure = createAction(
  DEPARTURE,
  ({ groupId, departureTime }) => ({
    groupId,
    departureTime,
  }),
);
export const fetchMyGroup = createAction(FETCH_MY_GROUP, groupId => groupId);
export const notifyDriverLocation = createAction(
  NOTIFY_DRIVER_LOCATION,
  ({ driverId, groupId, locationInfo }) => ({
    driverId,
    groupId,
    locationInfo,
  }),
);
export const confirmCost = createAction(
  CONFIRM_COST,
  ({ groupId, totalCost }) => ({
    groupId,
    totalCost,
  }),
);
export const initCostConfirmed = createAction(INIT_COST_CONFIRMED);

export const acceptGroupSaga = createRequestSaga(
  ACCEPT_GROUP,
  groupAPI.acceptGroup,
);
export const onTaxiSaga = createRequestSaga(ON_TAXI, groupAPI.onTaxi);
export const departureSaga = createRequestSaga(DEPARTURE, groupAPI.departure);
export const fetchMyGroupSaga = createRequestSaga(
  FETCH_MY_GROUP,
  groupAPI.fetchMyGroup,
);
export const notifyDriverLocationSaga = createRequestSaga(
  NOTIFY_DRIVER_LOCATION,
  groupAPI.notifyDriverLocation,
);
export const confirmCostSaga = createRequestSaga(
  CONFIRM_COST,
  groupAPI.confirmCost,
);

export function* groupSaga() {
  yield takeLatest(ACCEPT_GROUP, acceptGroupSaga);
  yield takeLatest(ON_TAXI, onTaxiSaga);
  yield takeLatest(DEPARTURE, departureSaga);
  yield takeLatest(FETCH_MY_GROUP, fetchMyGroupSaga);
  yield takeLatest(NOTIFY_DRIVER_LOCATION, notifyDriverLocationSaga);
  yield takeLatest(CONFIRM_COST, confirmCostSaga);
}

const initialState = {
  group: null,
  error: null,
  costConfirmed: false,
};

const group = handleActions(
  {
    [GROUP_CREATED]: (state, { payload: group }) => ({
      ...state,
      group,
    }),

    [GROUP_UPDATED]: (state, { payload: group }) => ({
      ...state,
      group,
    }),

    [UNLOAD_GROUP]: state => ({
      ...state,
      group: null,
    }),

    [GROUP_COST_UPDATED]: (state, { payload: group }) => ({
      ...state,
      group: {
        ...state.group,
        ...group,
      },
    }),

    [ACCEPT_GROUP_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),

    [ACCEPT_GROUP_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [ON_TAXI_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),

    [ON_TAXI_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [DEPARTURE_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),

    [DEPARTURE_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [FETCH_MY_GROUP_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),

    [FETCH_MY_GROUP_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [NOTIFY_DRIVER_LOCATION_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),
    [NOTIFY_DRIVER_LOCATION_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [CONFIRM_COST_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: {
        ...state.group,
        ...group,
      },
      error: null,
      costConfirmed: true,
    }),

    [CONFIRM_COST_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),
    [INIT_COST_CONFIRMED]: state => ({
      ...state,
      costConfirmed: initialState.costConfirmed,
    }),
  },
  initialState,
);

export default group;
