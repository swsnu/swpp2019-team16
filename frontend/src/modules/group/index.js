import group from './group';

export * from './group';
export default group;