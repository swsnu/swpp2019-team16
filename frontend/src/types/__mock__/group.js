import { createMockRider, mockDriver } from './user';

export const mockGroup = {
  groupId: 1,
  driver: mockDriver,
  riders: [
    createMockRider(1, 1, 'a@gmail.com'),
    createMockRider(2, 1, 'b@gmail.com'),
    createMockRider(3, 1, 'c@gmail.com'),
  ],
  fromLocation: 'A',
  toLocation: 'B',
};
