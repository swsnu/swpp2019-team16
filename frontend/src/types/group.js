import PropTypes from 'prop-types';
import { UserPropsTypes } from './user';

export const GroupPropTypes = PropTypes.shape({
  groupId: PropTypes.number.isRequired,
  driver: UserPropsTypes,
  riders: PropTypes.arrayOf(UserPropsTypes),
  from: PropTypes.string.isRequired,
  to: PropTypes.string.isRequired,
});
