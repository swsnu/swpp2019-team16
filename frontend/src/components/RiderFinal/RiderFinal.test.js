import React from 'react';
import { render } from '@testing-library/react';
import RiderFinal from './RiderFinal';
import { mockGroup } from 'types/__mock__/group';

describe('<RiderFinal />', () => {
  it('SHOULD match with snapshot', async () => {
    global.Math.floor = () => 1;
    const { container } = render(
      <RiderFinal group={mockGroup} onClickGoToMain={() => {}} />,
    );
    expect(container).toMatchSnapshot();
  });
});
