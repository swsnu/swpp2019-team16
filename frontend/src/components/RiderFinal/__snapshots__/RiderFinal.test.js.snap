// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RiderFinal /> SHOULD match with snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="makeStyles-whiteBox-3"
    >
      <div
        class="makeStyles-container-2"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          TotalCost is ₩
          100
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Your fee ₩
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          You saved ₩
          NaN
        </p>
        <div
          class="makeStyles-button-4"
          fullwidth="false"
          variant="contained"
        >
          Go To Main
        </div>
      </div>
    </div>
  </div>
</div>
`;
